### ISSUE TITLE
Introduce QueryConstraints to  LogicalPlan

### ISSUE THREAD

**Issue by mingmwang** [apache/datafusion#4222]
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
(This section helps Arrow developers understand the context and *why* for this feature, in addition to  the *what*)

I would like to introduce below structs to LogicalPlan and implement the QueryConstraints propagation logic.
Couple of logical optimization Rules(`FilterPushDown`, `ReduceOuterJoin`, `FilterNullJoinKeys`, `EliminateFilter`, `ReduceCrossJoin`...) can benefit from it and will simply the logic to deal with predicate/expr processing and will be able to cover more cases.

````
struct QueryConstraints {
    eq_props: EquivalenceProperties,
    null_props: NullProperties,
    not_null_props: NotNullProperties,
    other_constraints: Vec<Expr>,
    schema: SchemaRef
}
````
````
pub struct EquivalenceProperties {
    eq_classes: Vec<EquivalentClass>,
}
````

````
pub struct NullProperties {
    null_columns: Vec<Column>,
}
````

````
pub struct NotNullProperties {
    not_null_columns: Vec<Column>,
}
````
**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

**Comment by alamb** [comment]
Maybe this could be unified somehow with the analysis framework that @isidentical  is working on

**Comment by isidentical** [comment]
Exactly! The only major distinction I see is that, these seems to be constraints only for stuff that can be inferred at the logical planning time (so no stats, only schema + the existing query) which in turn might restrict some use cases but also might be much more reliable source. I am still in the middle on how we can unify these though 🤔

**Comment by alamb** [comment]
DataFusion now has https://docs.rs/datafusion/latest/datafusion/common/struct.Constraints.html which is what I think this PR proposes. Closing down
